jit-compiler/src/compiler.rs

```diff
@@ -0,0 +1,133 @@
+use libc::{c_void, dlopen, dlsym, RTLD_NOW};
+use mktemp::Temp;
+use std::{
+    collections::HashMap,
+    ffi::CString,
+    fs::{self},
+    process::Command,
+    str::from_utf8,
+};
+
+use powdr_ast::{
+    analyzed::Analyzed,
+    parsed::{
+        display::format_type_scheme_around_name,
+        types::{FunctionType, Type, TypeScheme},
+    },
+};
+use powdr_number::FieldElement;
+
+use crate::{codegen::escape_symbol, SymbolMap};
+
+// TODO make this depend on T
+
+const PREAMBLE: &str = r#"
+#![allow(unused_parens)]
+//type FieldElement = powdr_number::GoldilocksField;
+"#;
+
+pub fn generate_glue_code<T: FieldElement>(
+    symbols: &[&str],
+    analyzed: &Analyzed<T>,
+) -> Result<String, String> {
+    let mut glue = String::new();
+    let int_int_fun: TypeScheme = Type::Function(FunctionType {
+        params: vec![Type::Int],
+        value: Box::new(Type::Int),
+    })
+    .into();
+    for sym in symbols {
+        let ty = analyzed.type_of_symbol(sym);
+        if ty != int_int_fun && ty.ty != Type::Col {
+            return Err(format!(
+                "Only (int -> int) functions and columns are supported, but requested{}",
+                format_type_scheme_around_name(sym, &Some(ty)),
+            ));
+        }
+
+        // TODO we should use big int instead of u64
+        let name = escape_symbol(sym);
+        glue.push_str(&format!(
+            r#"
+            #[no_mangle]
+            pub extern "C" fn {}(i: u64) -> u64 {{
+                u64::try_from({name}(ibig::IBig::from(i))).unwrap()
+            }}
+            "#,
+            extern_symbol_name(sym)
+        ));
+    }
+
+    Ok(format!("{PREAMBLE}\n{glue}\n",))
+}
+
+const CARGO_TOML: &str = r#"
+[package]
+name = "powdr_jit_compiled"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+ibig = { version = "0.3.6", features = [], default-features = false }
+"#;
+
+/// Compiles the given code and returns the path to the
+/// temporary directory containing the compiled library
+/// and the path to the compiled library.
+pub fn call_cargo(code: &str) -> Result<(Temp, String), String> {
```

Comment:
Not a big deal, but you could make a `RustCompiler` struct that keeps the tmp dir and provides a `compile(&mut self, &str) -> Result<(), String>` and `lib_path(&self) -> String`. Then, the caller wouldn't have to know about the tmp dir.

But maybe that's overkill if the function is only `pub(crate)` and the only user is the `compile()` function, which would be the only function used outside this crate.